content/courses/intro-to-embedded-sys-prog/chapters/handling_interrupts.rst

```diff
@@ -6,177 +6,1537 @@ Handling Interrupts
+parameter. The association can also be achieved via a pragma with the
+same name as the aspect. Strictly speaking, the pragma
+:ada:`Attach_Handler` is obsolescent, but that just means that there is a
+newer way to make the association (i.e., the aspect). The pragma is not
+illegal and will remain supported. Because the pragma existed in a
+version of Ada prior to aspects, you will see a lot of existing code
+using the pragma. There's no language-driven reason to change the source
+code to use the aspect. New code should arguably use the aspect, but
+there's no technical reason to prefer one over the other.
+
+Here is an example of a protected object with one protected procedure
+interrupt handler. It uses the :ada:`Attach_Handler` aspect to tie a
+random number generator interrupt to the
+:ada:`RNG_Controller.Interrupt_Handler` procedure:
+
+.. code-block:: ada
+
+   protected RNG_Controller is
+      ...
+      entry Get_Random (Value : out UInt32);
+   private
+
+      Last_Sample    : UInt32 := 0;
+      Buffer         : Ring_Buffer;
+      Data_Available : Boolean := False;
+
+      procedure Interrupt_Handler with
+         Attach_Handler => Ada.Interrupts.Names.HASH_RNG_Interrupt;
+
+   end RNG_Controller;
+
+That's all that the developer must do to install the handler. The
+compiler and run-time library do the rest, automatically.
+
+The local variables are declared in the private part, as required by the
+language, because they are shared data meant to be protected from race
+conditions. Therefore, the only compile-time access possible is via visible
+subprograms and entries declare in the visible part. Those subprograms
+and entries execute with mutually exclusive access so no race conditions
+are possible.
+
+Note that procedure :ada:`Interrupt_Handler` is declared in the private
+part of :ada:`RNG_Controller`, rather than the visible part. That
+location is purely a matter of choice (unlike the variables), but there
+is a good reason to hide it: application software can call an interrupt
+handler procedure too. If you don't ever intend for that to happen, have
+the compiler enforce your intent. An alert code reader will know that
+only the hardware can call that procedure. If the handler is declared in
+the visible part, the reader must examine more of the code to determine
+whether there are any callers in the application code. Granted, a
+software call to an interrupt handler is rare, but not illegal, so you
+should state your intent in the code in an enforceable manner.
+
+Be aware that the Ada compiler is allowed to place restrictions on
+protected procedure handlers. The compiler can restrict the content of
+the procedure body, for example, or it might forbid calls to the handler
+from the application software. The rationale is to allow direct
+invocation by the hardware, to minimize interrupt latency to the extent
+possible.
+
+For completeness, here's the same :ada:`RNG_Controller` protected object
```

Comment:
Since this pragma is obsolescent, we could move the whole paragraph to an admonition, e.g.:

.. admonition:: For further information

    For completeness, here's the ...
```

Reply:
Same as above, I'd rather leave as-is.